esphome/components/teleinfo/teleinfo.cpp

```diff
@@ -0,0 +1,189 @@
+#include "teleinfo.h"
+#include "esphome/core/log.h"
+
+namespace esphome {
+namespace teleinfo {
+
+static const char *TAG = "teleinfo";
+
+/* Helpers */
+static int get_field(char *dest, char *buf_start, char *buf_end, int sep) {
+  char *field_end;
+  int len;
+
+  field_end = static_cast<char *>(memchr(buf_start, sep, buf_end - buf_start));
+  if (!field_end)
+    return 0;
+  len = field_end - buf_start;
+  strncpy(dest, buf_start, len);
+  dest[len] = '\0';
+
+  return len;
+}
+/* TeleInfo methods */
+bool TeleInfo::check_crc_(const char *grp, const char *grp_end) {
+  int grp_len = grp_end - grp;
+  uint8_t raw_crc = grp[grp_len - 1];
+  uint8_t crc_tmp = 0;
+  int i;
+
+  for (i = 0; i < grp_len - checksum_area_end_; i++)
+    crc_tmp += grp[i];
+
+  crc_tmp &= 0x3F;
+  crc_tmp += 0x20;
+  if (raw_crc != crc_tmp) {
+    ESP_LOGE(TAG, "bad crc: got %d except %d", raw_crc, crc_tmp);
+    return false;
+  }
+
+  return true;
+}
+bool TeleInfo::read_chars_until_(bool drop, uint8_t c) {
+  uint8_t received;
+  int j = 0;
+
+  while (available() > 0 && j < 128) {
+    j++;
+    received = read();
+    if (received == c)
+      return true;
+    if (drop)
+      continue;
+    /*
+     * Internal buffer is full, switch to OFF mode.
+     * Data will be retrieved on next update.
+     */
+    if (buf_index_ >= (MAX_BUF_SIZE - 1)) {
+      ESP_LOGE(TAG, "Internal buffer full");
```

Comment:
```suggestion
      ESP_LOGW(TAG, "Internal buffer full");
```

If it can be recovered from it should be a warning